Refactor address form component (properly this time)

chaimann opened the pull request:
Unblocks #6160 

## Summary
With the release of #6168 (addition of new fields to address form) previous updates to the address form component (#6191) were not enough for stock locations form in #6160 to work correctly, so, in order to fix that, this PR introduces fields composition to the address form component.

Break form into View Component slots, with more granular rendering control of each field. Form composition is decided at initialization, where user can provide `fields_preset` option (:contact by default) which then will pick relevant fields for given form. I.e. fields :name, :vat_id, :reverse_charge_status are relevant only for "contact" address type (e.g. customers, companies), while not relevant for "location" address type (e.g. stock locations).

Additionally, user can specify further which fields to include or exclude for even more customization, e.g. stock location model does not have `email` attribute yet, if we were to release new stock location form first and then migrate model to add `email` column, we can do that by specifying `exclude_columns: :email` when rendering stock location form and remove it when migration is in place.

### Showcase

https://github.com/user-attachments/assets/18838751-bcc1-4e7c-aaf9-cd2d003f086a


## Checklist

Check out our [PR guidelines](https://github.com/solidusio/.github/blob/master/CONTRIBUTING.md#pull-request-guidelines) for more details.

The following are mandatory for all PRs:

- [x] [I agree that my PR will be published under the same license as Solidus](https://github.com/solidusio/solidus/blob/main/LICENSE.md).
- [x] I have written a thorough PR description.
- [x] I have kept my commits small and atomic.
- [x] [I have used clear, explanatory commit messages](https://github.com/solidusio/.github/blob/main/CONTRIBUTING.md#writing-good-commit-messages).

The following are not always needed:

- 📖 I have updated the README to account for my changes.
- 📑 I have documented new code [with YARD](https://www.rubydoc.info/gems/yard/file/docs/Tags.md).
- 🛣️ I have opened a PR to update the [guides](https://github.com/solidusio/edgeguides).
- ✅ I have added automated tests to cover my changes.
- 📸 I have attached screenshots to demo visual changes.

tvdeyen commented on admin/app/components/solidus_admin/ui/forms/address/component.rb:
```diff
@@ -1,15 +1,83 @@
 
 class SolidusAdmin::UI::Forms::Address::Component < SolidusAdmin::BaseComponent
-  def initialize(addressable:, form_field_name:, disabled: false, include_name_field: true)
+  DEFAULT_FIELDS = %i[
```
I think if we make these two constants private methods one could inherit from this component and override this in order to customize its one address form more easily.

jarednorman replied:
I'd love for us to favour composition-based extension approaches rather than inheritance-based ones. I know that SolidusAdmin is pre-1.0, but I've already seen the pain of changes to components breaking my subclasses. I think in this case we should _not_ provide the subclass option and just let people wrap the class if they want.

tvdeyen reviewed (commented):
Love it. I have one note about easy customization of this component

tvdeyen commented:
Agreed. I just don't think constants are the right way to store sets of fields. Do you have an example where constants help with composition?

jarednorman commented:
I don't think it necessarily encourages composition, but I would say it encourages subclassing. If we want to make that functionality controllable, perhaps a different approach would be better... but I'm not sure exactly how that would all fit together.

chaimann commented:
hi guys! If by customizing you mean allowing to include completely custom fields along with the default ones provided (or skip defaults at all), it's possible to do so without the need to subclass/override current component, I'll just need to make a few changes to the component.

tvdeyen commented on admin/lib/solidus_admin/ext/view_component/slotable_default.rb:
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
```
if  we move the file somewhere into app/** it would be auto loaded by Rails's autoloader. 

Since this is a module anyway, why not put it into `app/components/concerns` to match Rails' default module paths?

chaimann replied:
the reason why I decided not to put it into components concerns folder is that it's a monkey-patch of [ViewComponent::SlotableDefault](https://github.com/ViewComponent/view_component/blob/main/lib/view_component/slotable_default.rb) and I usually like to place monkey-patches/extensions separately from other classes/modules (hence the `ext` folder), how about we place it in `app/lib/ext`? so that it is still separate and we make use of Rails autoload

tvdeyen replied:
I understood that this is a monkey patch. There is already a place in solidus (`app/decorators`) for them, but since you explicitly include this anyway, we can simply use the concerns folder. I order to not forget about this file it is good practice to build a version guard into your monkey patch

```rb 
if Gem::Version.new("3.22.0") < Gem::Version.new(ViewComponent::Version)
  raise "remove me"
end
```

chaimann replied:
good to know thanks!
just out of interest, I can't find `app/decorators` folder anywhere

tvdeyen replied:
Not in core, true. But all extensions and many apps use this pattern. 

Thats also a reason why I dont want to introduce it and make use of the concerns folder.

tvdeyen commented on admin/lib/solidus_admin/engine.rb:
```diff
@@ -69,5 +69,11 @@ class Engine < ::Rails::Engine
     end
+
+    initializer "solidus_admin.load_ext" do
+      Dir[root.join("lib/solidus_admin/ext/**/*.rb")].sort.each do |file|
```
This seems to be another way of loading concerns/monkeypatches. I am not sure we need that if we make use of Rails autoloading

tvdeyen commented on admin/app/components/solidus_admin/ui/forms/address/fields/reverse_charge_fields/component.html.erb:
```diff
@@ -0,0 +1,9 @@
+<% if Spree::Backend::Config.show_reverse_charge_fields %>
```
I am not a huge friend of having surprising behaviour like this in a template. We could make use of [the `render?` callback](https://viewcomponent.org/api.html#render--boolean) in the component class instead.

tvdeyen reviewed (commented):
Nice idea. I need to play around with it a bit more but it looks like a great way of adding address components. I have some nits about placing the view component extension module.

chaimann commented:
@tvdeyen I've addressed your comments, and, as I mentioned in one of the commits, the issue with ViewComponent default_slot has been fixed, but it won't be around until v4, so the patch is now using an updated solution from their PR and a guard has been added to remove the patch once we have upgraded VC to v4
